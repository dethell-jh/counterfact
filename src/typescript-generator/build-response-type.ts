type OmitByValue<Base, Condition> = Pick<
  Base,
  {
    [Key in keyof Base]: [Base[Key]] extends [Condition] ? never : Key;
  }[keyof Base]
>;

type StatusCodes = "200_OK" | "404_NOT_FOUND" | "500_INTERNAL_SERVER_ERROR";

interface CounterfactResponse {
  status?: StatusCodes;
  contentType: string;
  body?: unknown;
  headers?: { [key: string]: number | string };
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Anything = any;

interface AnyResponseWithContentType<ContentType> {
  contentType: ContentType;
  status: Anything;
  body: Anything;
  headers: Anything;
}

type MatchFunction<
  AllResponses extends CounterfactResponse,
  RemainingResponses extends AllResponses = AllResponses
> = <ContentType extends RemainingResponses["contentType"]>(
  contentType: ContentType,
  body: Extract<
    RemainingResponses,
    AnyResponseWithContentType<ContentType>
  >["body"],
  headers?: Extract<
    RemainingResponses,
    AnyResponseWithContentType<ContentType>
  >["headers"]
) => BuildResponseTypeForStatusCode<
  AllResponses,
  Exclude<RemainingResponses, AnyResponseWithContentType<ContentType>>
>;

type IfResponsesHaveContentType<Responses, ContentType, Result> =
  AnyResponseWithContentType<ContentType> extends Responses ? Result : never;

type ShortcutFunction<
  AllResponses extends CounterfactResponse,
  RemainingResponses extends AllResponses = AllResponses,
  ContentType extends RemainingResponses["contentType"] = AllResponses["contentType"]
> = (
  body: Extract<RemainingResponses, { contentType: ContentType }>["body"],
  headers?: Extract<
    RemainingResponses,
    AnyResponseWithContentType<ContentType>
  >["headers"]
) => BuildResponseTypeForStatusCode<
  AllResponses,
  Exclude<RemainingResponses, AnyResponseWithContentType<ContentType>>
>;

type MaybeShortcutFunction<
  AllResponses extends CounterfactResponse,
  RemainingResponses extends AllResponses,
  ContentType extends RemainingResponses["contentType"] = AllResponses["contentType"]
> = IfResponsesHaveContentType<
  RemainingResponses,
  ContentType,
  ShortcutFunction<AllResponses, RemainingResponses, ContentType>
>;

type BuildResponseTypeForStatusCode<
  AllResponses extends CounterfactResponse,
  RemainingResponses extends AllResponses = AllResponses
> = [RemainingResponses] extends [never]
  ? AllResponses
  : OmitByValue<
      {
        match: MatchFunction<AllResponses, RemainingResponses>;
        json: MaybeShortcutFunction<
          AllResponses,
          RemainingResponses,
          "application/json"
        >;
        text: MaybeShortcutFunction<
          AllResponses,
          RemainingResponses,
          "text/plain"
        >;
        html: MaybeShortcutFunction<
          AllResponses,
          RemainingResponses,
          "text/html"
        >;
      },
      never
    >;

// FlattResponse isn't quite right but close enough for now. It should return something that looks like:
// { status1, contentType1, headers1, body1 } | { status2, contentType2, headers2, body2 } | { status3, contentType3, headers3, body3 }
// It currently returns:
// { contentType1,  body1 } | { contentType2,  body2 } | { contentType3, headers3 } & { headers1 | headers2 | headers3 , status1 | status2 | status3 }

type FlattenResponses<
  Responses extends {
    [contentType: string]: {
      headers: number | { [name: string]: string };
      content: {
        contentType: string;
        body: unknown;
      };
    };
  }
> = Responses[keyof Responses]["content"] & {
  headers: Responses[keyof Responses]["headers"];
  status: StatusCodes;
};

// eslint-disable-next-line import/exports-last
export type BuildResponseType<
  ResponsesForStatusCode extends {
    [key: string]: {
      content: CounterfactResponse;
      headers?: { [key: string]: number | string };
    };
  }
> = {
  [StatusCode in keyof ResponsesForStatusCode]: BuildResponseTypeForStatusCode<
    ResponsesForStatusCode[StatusCode]["content"] & {
      headers: ResponsesForStatusCode[StatusCode]["headers"];
      status: StatusCode;
    }
  >;
};

/* ************************************************************************ */
/* ************************************************************************ */
/* ************************************************************************ */
// This goes in the generated path-types/**/*.types.ts file
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
type Responses = {
  "200_OK": {
    headers: {
      "x-custom-header": string;
    };
    content:
      | {
          contentType: "text/plain";
          body: string;
        }
      | { contentType: "application/json"; body: { message: string } }
      | { contentType: "text/html"; body: string };
  };
  "404_NOT_FOUND": {
    headers: {
      "x-404-header": string;
    };
    content:
      | {
          contentType: "text/plain";
          body: string;
        }
      | { contentType: "application/json"; body: { message: string } }
      | { contentType: "text/html"; body: string };
  };
  DEFAULT: {
    headers: {
      "x-default-header": string;
    };
    content:
      | {
          contentType: "text/plain";
          body: string;
        }
      | { contentType: "text/html"; body: string };
  };
};

export type HTTP_GET = ({
  response,
  context,
}: {
  response: BuildResponseType<Omit<Responses, "DEFAULT">>;
  context: { found: () => boolean; message: () => "Hello World" };
}) => FlattenResponses<Responses>;

/* ************************************************************************ */
/* ************************************************************************ */
/* ************************************************************************ */

// How it's used in the paths/**/*.ts file
export const GET: HTTP_GET = ({ response, context }) => {
  if (!context.found()) {
    return response["404_NOT_FOUND"]
      .match("text/plain", "not found", { "x-404-header": "bar" })
      .match("text/html", "<h1>not found</h1>")
      .match("application/json", { message: "not found" });
  }

  return response["200_OK"]
    .text(context.message(), { "x-custom-header": "bar" })
    .html(`<h1>${context.message()}</h1>`)
    .json({ message: context.message() });
};
