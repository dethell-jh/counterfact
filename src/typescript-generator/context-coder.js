import nodePath from "node:path";

import { Coder } from "./coder.js";

export class ContextCoder extends Coder {
  isRoot() {
    return this.requirement.url.split("/").at(-2).split("~1").length === 2;
  }

  names() {
    return super.names("Context");
  }

  write() {
    if (this.isRoot()) {
      return "{}";
    }

    return { raw: 'export { default } from "../$context.js"' };
  }

  modulePath() {
    const pathString = this.requirement.url
      .split("/")
      .at(-2)
      .replaceAll("~1", "/");

    return `${nodePath.join(
      "paths",
      nodePath.dirname(pathString)
    )}/$context.ts`;
  }
}
