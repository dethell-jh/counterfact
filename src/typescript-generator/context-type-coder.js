import nodePath from "node:path";

import { Coder } from "./coder.js";

export class ContextTypeCoder extends Coder {
  constructor(requirement, isRoot) {
    super(requirement);
    this.isRoot = isRoot;
  }

  pathString() {
    return this.requirement.url
      .split("/")
      .at(-2)
      .replaceAll("~1", "/")
      .replaceAll("~0", "~");
  }

  get id() {
    return "ContextTypeCoder";
  }

  names() {
    return super.names("ContextType");
  }

  write(script) {
    if (script.path === "paths/$.context.ts") {
      return "typeof Context";
    }

    return { raw: 'export type { ContextType } from "../$.context"' };
  }

  modulePath() {
    return nodePath
      .join("paths", nodePath.dirname(this.pathString()), "$.context.ts")
      .replaceAll("\\", "/");
  }
}
