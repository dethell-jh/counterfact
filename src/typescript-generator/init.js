import nodePath from "node:path";
import { fileURLToPath } from "node:url";

import yaml from "js-yaml";
import fs from "fs-extra";

const rootPath = nodePath.join(
  nodePath.dirname(fileURLToPath(import.meta.url)),
  "../../"
);

export async function init(source, destination) {
  // eslint-disable-next-line node/no-sync
  if (fs.existsSync(destination)) {
    throw new Error(`Destination already exists: ${destination}`);
  }

  await fs.copy(nodePath.join(rootPath, "templates/typescript"), destination);

  const specification = await yaml.load(await fs.readFile(source, "utf8"));

  specification.servers.unshift({ url: "/" });

  await fs.mkdir(nodePath.join(destination, "counterfact", "openapi"));

  await fs.writeFile(
    nodePath.join(destination, "counterfact", "openapi", "openapi.yaml"),
    yaml.dump(specification)
  );
}
