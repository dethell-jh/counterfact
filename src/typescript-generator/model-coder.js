import nodePath from "node:path";

import { Coder } from "./coder.js";

export class ModelCoder extends Coder {
  names() {
    return super.names("Model");
  }

  write() {
    return "class {}";
  }

  modulePath() {
    const pathString = this.requirement.url
      .split("/")
      .at(-2)
      .replaceAll("~1", "/");

    return `${nodePath.join("paths", nodePath.dirname(pathString))}/#model.ts`;
  }
}
