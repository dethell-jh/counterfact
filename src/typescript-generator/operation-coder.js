import nodePath from "node:path";

import { Coder } from "./coder.js";
import { SchemaCoder } from "./schema-coder.js";
import { OperationTypeCoder } from "./operation-type-coder.js";
import { SchemaTypeCoder } from "./schema-type-coder.js";

export class OperationCoder extends Coder {
  requestMethod() {
    return this.requirement.url.split("/").at(-1).toUpperCase();
  }

  names() {
    return super.names(this.requestMethod());
  }

  write(script) {
    const responses = this.requirement.select("responses");

    const returns = responses.map(([statusCode, response]) => {
      if (statusCode === "default") {
        return this.responseForStatusCode(script, response);
      }

      return `if (statusCode === "${statusCode}") { 
        ${this.responseForStatusCode(script, response, statusCode)}
      }`;
    });

    const statusCodes = responses.map(([statusCode]) => statusCode);

    const requestProperties = this.requirement.data.parameters
      ? Array.from(
          new Set(
            this.requirement.data.parameters.map((parameter) => parameter.in)
          )
        )
      : [];

    if (this.requestMethod() !== "GET") {
      requestProperties.push("body");
    }

    requestProperties.push("context", "tools");

    return `({ ${requestProperties.join(", ")} }) => {
      const statusCode = tools.oneOf(${JSON.stringify(statusCodes)});

      ${returns.join("\n")}
 
      return {
        status: 415,
        contentType: "text/plain",
        body: "HTTP 415: Unsupported Media Type",
      }
    }`;
  }

  responseForStatusCode(script, response, statusCode) {
    const returns = response
      .select("content")
      .map(([contentType, responseForContentType]) =>
        this.responseForContentType(
          statusCode,
          contentType,
          responseForContentType,
          script
        )
      );

    return returns.join("\n");
  }

  // eslint-disable-next-line max-params
  responseForContentType(statusCode, contentType, response, script) {
    const statusLine = statusCode === undefined ? "" : `status: ${statusCode},`;

    const exampleKeys = Object.keys(response.select("examples")?.data ?? {});
    const examples = (response.select("examples") || []).map(
      ([name, example]) => `if (example === "${name}") { 
          
            return {
              ${statusLine}
              contentType: "${contentType}",
              body: ${JSON.stringify(example.select("value").data)}
            }
        }`
    );

    const bodyType = new SchemaTypeCoder(response.select("schema")).write(
      script
    );

    const exampleSelector =
      examples.length === 0
        ? ""
        : `const example = tools.oneOf(${JSON.stringify(exampleKeys)});`;

    return `if (tools.accepts("${contentType}")) { 
      ${exampleSelector}

      ${examples.join("\n")}

 

      return {
        ${statusLine}
        contentType: "${contentType}",
        body: tools.randomFromSchema(${new SchemaCoder(
          response.select("schema")
        ).write(script)}) as ${bodyType}
      };
    }`;
  }

  typeDeclaration(namespace, script) {
    const operationTypeCoder = new OperationTypeCoder(this.requirement);

    return script.importType(operationTypeCoder);
  }

  modulePath() {
    const pathString = this.requirement.url
      .split("/")
      .at(-2)
      .replaceAll("~1", "/");

    return `${nodePath.join("path", pathString)}.types.ts`;
  }
}
