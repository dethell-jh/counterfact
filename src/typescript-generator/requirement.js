export class Requirement {
  constructor(data, url = "", specification = undefined) {
    if (data === undefined) {
      throw new Error("cannot create requirement without data");
    }

    this.data = data;
    this.url = url;
    this.specification = specification;
  }

  get isReference() {
    return this.data?.$ref !== undefined;
  }

  reference() {
    return this.specification.requirementAt(this.data.$ref, this.url);
  }

  has(item) {
    return item in this.data;
  }

  get(item) {
    if (!this.has(item)) {
      return undefined;
    }

    return new Requirement(
      this.data[item],
      `${this.url}/${item}`,
      this.specification
    );
  }

  select(path, data = this.data, basePath = "") {
    const [head, ...tail] = path.split("/");

    const branch = data[this.unescapeJsonPointer(head)];

    if (!branch) {
      return undefined;
    }

    if (tail.length === 0) {
      return new Requirement(
        branch,
        `${this.url}/${basePath}${head}`,
        this.specification
      );
    }

    return this.select(tail.join("/"), branch, `${basePath}${head}/`);
  }

  toJSON() {
    return {
      specification: this.specification,
      url: this.url,
      data: this.data,
    };
  }

  forEach(callback) {
    Object.keys(this.data).forEach((key) => {
      callback([key, this.select(this.escapeJsonPointer(key))]);
    });
  }

  map(callback) {
    const result = [];

    // eslint-disable-next-line array-callback-return
    this.forEach(([key, value]) => result.push(callback(value, key)));

    return result;
  }

  flatMap(callback) {
    const result = [];

    // eslint-disable-next-line array-callback-return
    this.forEach(([key, value]) => result.push(callback(value, key)));

    return result.flat();
  }

  escapeJsonPointer(string) {
    return string.replaceAll("~", "~0").replaceAll("/", "~1");
  }

  unescapeJsonPointer(pointer) {
    return pointer.replaceAll("~1", "/").replaceAll("~0", "~");
  }
}
