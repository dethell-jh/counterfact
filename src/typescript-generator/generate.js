import fs from "node:fs/promises";
import nodePath from "node:path";

import { Repository } from "./repository.js";
import { Specification } from "./specification.js";
import { OperationCoder } from "./operation-coder.js";

// eslint-disable-next-line no-underscore-dangle
const __dirname = nodePath.dirname(new URL(import.meta.url).pathname);

export async function generate(
  source,
  destination,
  repository = new Repository()
) {
  const specification = new Specification();

  const requirement = await specification.requirementAt(`${source}#/paths`);

  requirement.forEach((pathDefinition, key) => {
    pathDefinition.forEach((operation) => {
      repository.get(`paths${key}.ts`).export(new OperationCoder(operation));
    });
  });

  await repository.writeFiles(destination);

  fs.copyFile(
    nodePath.join(__dirname, "../../templates/response-builder-factory.ts"),
    nodePath.join(destination, "response-builder-factory.ts")
  );
}
