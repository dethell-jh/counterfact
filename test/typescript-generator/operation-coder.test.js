import prettier from "prettier";

import { OperationCoder } from "../../src/typescript-generator/operation-coder";
import { Requirement } from "../../src/typescript-generator/requirement.js";

function format(code) {
  return prettier.format(code, { parser: "typescript" });
}

describe("an OperationCoder", () => {
  it("generates a list of potential names", () => {
    const coder = new OperationCoder(new Requirement({}, "#/paths/hello/get"));

    const [one, two, three] = coder.names();

    expect([one, two, three]).toStrictEqual(["GET", "GET2", "GET3"]);
  });
});
