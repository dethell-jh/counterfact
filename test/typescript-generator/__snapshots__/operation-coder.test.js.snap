// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`an OperationCoder generates a complex get operation 1`] = `
"({ path, query, context, tools }) => {
  const statusCode = tools.oneOf([\\"200\\", \\"400\\", \\"default\\"]);

  if (statusCode === \\"200\\") {
    if (tools.accepts(\\"application/json\\")) {
      const example = tools.oneOf([\\"first-example\\", \\"second-example\\"]);

      if (example === \\"first-example\\") {
        return {
          status: 200,
          contentType: \\"application/json\\",
          body: \\"first\\",
        };
      }
      if (example === \\"second-example\\") {
        return {
          status: 200,
          contentType: \\"application/json\\",
          body: \\"second\\",
        };
      }

      return {
        status: 200,
        contentType: \\"application/json\\",
        body: tools.randomFromSchema(schema) as Type,
      };
    }
  }
  if (statusCode === \\"400\\") {
    if (tools.accepts(\\"application/json\\")) {
      return {
        status: 400,
        contentType: \\"application/json\\",
        body: tools.randomFromSchema(schema) as Type,
      };
    }
  }
  if (tools.accepts(\\"application/json\\")) {
    return {
      contentType: \\"application/json\\",
      body: tools.randomFromSchema(schema) as Type,
    };
  }

  return {
    status: 415,
    contentType: \\"text/plain\\",
    body: \\"HTTP 415: Unsupported Media Type\\",
  };
};
"
`;

exports[`an OperationCoder generates a simple post operation 1`] = `
"({ body, context, tools }) => {
  const statusCode = tools.oneOf([\\"default\\"]);

  if (tools.accepts(\\"application/json\\")) {
    return {
      contentType: \\"application/json\\",
      body: tools.randomFromSchema(schema) as Type,
    };
  }

  return {
    status: 415,
    contentType: \\"text/plain\\",
    body: \\"HTTP 415: Unsupported Media Type\\",
  };
};
"
`;
